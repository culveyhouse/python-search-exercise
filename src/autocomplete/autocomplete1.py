import re

class Autocomplete1:
    def __init__(self, suggestions):
        self.suggestions = suggestions

    def performSearch(self, term, limit=None):
        """Searches through the possible suggestions, creating an array of
        matches for any books whose title matches the specified search term.

        This should perform a case sensitive contains search.

        Example:
            With 'biology' specified as the term a book with the title
            'Microbiology' should be returned. A book with the title 'Biology'
            should not be returned.

        Args:
            term = the term to search for in book titles.
            limit = maximum number of autocomplete suggestions to return 

        Return:
            List of results (dictionary entries) where the title contains 
            the search term.
        """
        
        # A list of dictionary entries that match the "term" argument
        results = [] 

        # Compiled regex search using "term" argument for efficiency
        term_search = re.compile(r"{}".format(term)) 

        for suggestion in self.suggestions:
            # Our suggestions may contain special UTF-8 characters            
            title = suggestion['title'].encode("utf-8")
            if term_search.search(title): 
                results.append(suggestion)
                
                # Limit autocomplete results if limit passed
                if limit and len(results)>=limit: 
                    break

        return results
