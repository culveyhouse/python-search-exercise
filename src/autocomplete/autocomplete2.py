import re

class Autocomplete2:
    def __init__(self, suggestions):
        self.suggestions = suggestions

    def performSearch(self, term, limit=None):
        """Searches through the possible suggestions, creating an array of
        matches for any books whose title matches the specified search term.

        This should perform a case sensitive prefix search.

        Example:
            With 'Biology' specified as the term a book with the title
            'Biology' should be returned. A book with the title
            'Microbiology' should not be returned.

        Args:
            term - the term to search for in book titles.
            limit = maximum number of autocomplete suggestions to return            

        Return:
            List of results (dictionary entries) where the title contains 
            the search term.
        """
        
        # A list of dictionary entries that match the "term" argument
        results = [] 

        # Compiled regex search using "term" argument for efficiency
        prefix_search = re.compile(r"^{}".format(term)) 

        for suggestion in self.suggestions:
            # Our suggestions may contain special UTF-8 characters
            title = suggestion['title'].encode("utf-8")            
            if prefix_search.match(title): 
                results.append(suggestion)
                
                # Limit autocomplete results if limit passed
                if limit and len(results)>=limit: 
                    break

        return results
