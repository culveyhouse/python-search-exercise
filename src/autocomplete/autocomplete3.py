import re

class Autocomplete3:
    def __init__(self, suggestions):
        self.suggestions = suggestions

    def performSearch(self, terms, limit=None, offset=None, fields=None):
        """Searches through the possible suggestions, creating an array of
        matches for any books whose title matches the specified search term.

        This method performs a case insensitive contains search and sort the
        results by title. Additionally, it takes two optional parameters for
        paginated results, limit and offset. Limit (also called count in 
        other classes) is the maximum number of results to return, and offset 
        denotes the number of records to pop from the sorted list of 
        dictionaries from the start of the list. 

        Example:
            With 'BIOLOGY' specified as the term a book with the title
            'Biology' should be returned. A book with the title 'Microbiology'
            should also be returned.

        Args:
            term = the term to search for in book titles.
            limit = maximum number of autocomplete suggestions to return 
            offset = number of dictionary entries to pop/ignore from list start
            (Note: limit and offset must be passed together, or neither)
            fields = an optional subset of dictionary keys to return for each result

        Return:
            List of results (dictionary entries) where the title contains 
            the search term.
        """

        # If limit or offset is passed, both must be present
        if (offset is None) is not (limit is None):
            raise Exception("Autocomplete3.performSearch Error: You must pass " 
            "an offset and limit, or neither")
            
        # A list of dictionary entries that match the "term" argument
        results = [] 
        term_search = {}
        if type(terms) is str:
            terms = {'title':str(terms)}

        # Compile a case-insensitive regex search for each search term
        for k, v in terms.iteritems():
            terms[k] = str(v)
            term_search[k] = re.compile(r"{}".format(v), re.I)

        # Boolean to return whether there are more paginated results beyond limit
        truncated = False

        for suggestion in self.suggestions:
            # Our suggestions may contain special UTF-8 characters
            match_fail = False
            for key, search in term_search.iteritems():
                suggestion_value = repr(suggestion[key]).encode("utf-8")
                if not term_search[key].search(suggestion_value):
                    match_fail = True
                    continue
            if not match_fail:
                results.append(suggestion)

        sorted_results = sorted(results, key = lambda suggestion: suggestion['title'])

        # Handle pagination / truncation if limit & offset were passed
        if offset>=0 and limit:
            # Check if remaining entries after offset are more than max per page
            if len(sorted_results[offset:])>limit:
                truncated = True
            # The transformed slice of sorted results for the current page
            sorted_results = sorted_results[offset:(offset+limit)]

        # Return only a subset of the total fields of each dict entry if specified
        if fields:
            subset_results = []
            for entry in sorted_results:
                this_subset = {}
                for k,v in entry.iteritems():
                    if k in fields:
                        this_subset[k] = v
                subset_results.append(this_subset)
            sorted_results = subset_results
            del subset_results
        
        """Return only list of sorted_results unless the list was truncated by 
        limit/offset arguments, in which case indicate whether there was a 
        remainder that was truncated 
        """
        return (sorted_results, truncated) if offset>=0 else sorted_results  
        