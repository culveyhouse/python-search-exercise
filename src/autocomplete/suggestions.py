import json
import os

SUGGESTIONS_PATH = os.path.join(
    os.path.dirname(__file__), '../../data/suggestions.json'
)

class Suggestions:
    @classmethod
    def load(cls):
        with open(SUGGESTIONS_PATH, 'r') as f:
            return json.load(f)['data']
