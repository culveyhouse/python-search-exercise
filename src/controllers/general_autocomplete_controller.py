import urllib

# Required since Pytest and simple HTTP server do not handle relative paths the same
import sys, os
sys.path.append(os.path.realpath(os.path.dirname(__file__)+"/.."))

from autocomplete.autocomplete3 import Autocomplete3
from autocomplete.suggestions import Suggestions

class GeneralAutocompleteController:
    @classmethod
    def index(cls, query, fields, count, pageStart):
        """Performs a case insensitive contains search on the given fields,
        optionally returning a subset of the suggestion objects in pages.

        Example:
            http://localhost:9000/generalindex?q=title:biology,attribution:raven&fields=title,subtitle,edition&count=20&pageStart=<pageKey>
        
        Args:
            query - the search query. This is a string of comma seperated
                    key/value pairs to filter the search on. For example, this
                    query string, 'title:biology,attribution:raven' should
                    search for books where the title contains 'biology' and
                    where the attribution field contains 'raven'.
            fields - the fields that should be returned. This is a string
                     of comma seperated fields(e.g. 'title,subtitle,edition')
            count -  the max number of results to return.
            pageStart - parameter to indicate where to start the page.

        Return:
            Dictionary containing the list of results that match the searchTerm
            and contain only the fields specified. Paging information should
            also be included in the dictionary. Return dictionary should
            include these fields in this format:

            {
                'status': {
                    'moreResults': <Boolean>
                    'pageNext': <String|Int>
                },
                'result': <List>
            }
        """

        search_dict = {} # dict to store the converted key:value search terms
        more_results = False # Boolean to indicate whether there are more results
        pageNext = None # Int to indicate the paging key for use with next page link
        if not pageStart:
            pageStart = 1 # E stablish human-readable "page 1" if no pagination yet

        # Decode any URL encoded characters in the searchTerm query string
        search_kv = query.split(',')
        for kv in search_kv:
            k,v = kv.split(':')
            v = urllib.unquote(v)
            search_dict[k]=v
        
        if fields:
            fields = fields.split(',')

        data = Suggestions.load() # load JSON data from data/suggestions.json
        
        # Autocomplete3 has all functionality necessary for this controller
        autocomplete = Autocomplete3(data) 
        
        """Get list of dictionary items up to max "count", as well as the 
        boolean moreResults which indicates whether more pages of results exist
        """
        results, more_results = autocomplete.performSearch(search_dict, count, 
            (pageStart-1)*count, fields)

        status = {'moreResults': more_results}
        
        # Add "pageNext" to status dict if the performSearch indicated more results
        if more_results: 
            status['pageNext'] = pageStart+1
        
        return_dict = {
            'status': status,
            'result': results
        }

        return return_dict
