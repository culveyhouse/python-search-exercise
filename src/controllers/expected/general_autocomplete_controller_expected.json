[
    {
        "copyrightYear": 2012, 
        "edition": "12th edition", 
        "primaryCategory": "Business & Finance", 
        "title": "Consumer Behavior"
    }, 
    {   
        "copyrightYear": 2012, 
        "edition": "14th edition", 
        "primaryCategory": "Business & Finance", 
        "title": "Contemporary Advertising and Integrated Marketing Communications"
    }, 
    {   
        "copyrightYear": 2012,         
        "edition": "10th edition", 
        "primaryCategory": "Business & Finance", 
        "title": "Corporate Finance"
    }, 
    {   
        "copyrightYear": 2012, 
        "edition": "5th edition", 
        "primaryCategory": "Business & Finance", 
        "title": "Essentials of Contemporary Management"
    }, 
    {   
        "copyrightYear": 2012, 
        "edition": "9th edition", 
        "copyrightYear": 2012, 
        "primaryCategory": "Business & Finance", 
        "title": "Essentials of Investments"
    }
]
